src/scale_impls/encode.rs

```diff
@@ -80,82 +83,136 @@ impl<T> EncodeAsFields for Composite<T> {
 // can't handle encoding to sequences/arrays. However, we can encode safely into sequences here because we can inspect the
 // values we have and more safely skip newtype wrappers without also skipping through types that might represent 1-value
 // sequences/arrays for instance.
-fn encode_composite<T>(
+fn encode_composite<'a, T, R: TypeResolver>(
     value: &Composite<T>,
-    mut type_id: u32,
-    types: &PortableRegistry,
+    mut type_id: &'a R::TypeId,
+    types: &'a R,
     out: &mut Vec<u8>,
 ) -> Result<(), Error> {
     // Encode our composite Value as-is (pretty much; we will try to
     // unwrap the Value only if we need primitives).
-    fn do_encode_composite<T>(
+    fn do_encode_composite<T, R: TypeResolver>(
         value: &Composite<T>,
-        type_id: u32,
-        types: &PortableRegistry,
+        type_id: &R::TypeId,
+        types: &R,
         out: &mut Vec<u8>,
     ) -> Result<(), Error> {
-        let ty =
-            types.resolve(type_id).ok_or_else(|| Error::new(ErrorKind::TypeNotFound(type_id)))?;
-        match &ty.type_def {
-            TypeDef::Tuple(_) | TypeDef::Composite(_) => match value {
-                Composite::Named(vals) => {
-                    let keyvals =
-                        vals.iter().map(|(key, val)| (Some(&**key), val as &dyn EncodeAsType));
-                    EncodeComposite(keyvals).encode_as_type_to(type_id, types, out)
+        struct EncodingVisitor<'a, T, R: TypeResolver> {
```

Comment:
Note: Using the closure based `ConcreteResolvedTypeVisitor` would be a bit simpler here, but it does not work because all the closures would need to take a mutable reference to `out: &mut Vec<u8>` at the same time.

Reply:
`scale_type_resolver::visitor::new` takes a context as a first argument for exactly this reason; just pass the mutable vec as the context and it'll be handed to whichever closure is called :)